src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs

```diff
@@ -228,16 +244,67 @@ public AnalyzerOptions AnalyzerOptions
+        private sealed class WorkspaceAnalyzerConfigOptionsProvider : AnalyzerConfigOptionsProvider
         {
-            return new AnalyzerOptions(additionalDocStates.Values.Select(d => new AdditionalTextDocument(d)).ToImmutableArray<AdditionalText>());
+            private readonly ProjectState _projectState;
+
+            public WorkspaceAnalyzerConfigOptionsProvider(ProjectState projectState)
+            {
+                _projectState = projectState;
+            }
+
+            public override AnalyzerConfigOptions GetOptions(SyntaxTree tree)
+            {
+                return new WorkspaceAnalyzerConfigOptions(_projectState._lazyAnalyzerConfigSet.GetValue(CancellationToken.None).GetOptionsForSourcePath(tree.FilePath));
+            }
+
+            public override AnalyzerConfigOptions GetOptions(AdditionalText textFile)
+            {
+                // TODO: correctly find the file path, since it looks like we give this the document's .Name under the covers if we don't have one
+                return new WorkspaceAnalyzerConfigOptions(_projectState._lazyAnalyzerConfigSet.GetValue(CancellationToken.None).GetOptionsForSourcePath(textFile.Path));
+            }
+
+            // PROTOTYPE: why isn't this just a provided implementation?
+            private sealed class WorkspaceAnalyzerConfigOptions : AnalyzerConfigOptions
+            {
+                private readonly ImmutableDictionary<string, string> _analyzerOptions;
+
+                public WorkspaceAnalyzerConfigOptions(AnalyzerConfigOptionsResult analyzerConfigOptions)
+                {
+                    _analyzerOptions = analyzerConfigOptions.AnalyzerOptions;
+                }
+
+                public override bool TryGetValue(string key, out string value) => _analyzerOptions.TryGetValue(key, out value);
+            }
+        }
+
+        private static ValueSource<AnalyzerConfigSet> ComputeAnalyzerConfigSetValueSource(IEnumerable<AnalyzerConfigDocumentState> analyzerConfigDocumentStates)
+        {
+            return new AsyncLazy<AnalyzerConfigSet>(
+                asynchronousComputeFunction: async cancellationToken =>
+                {
+                    var tasks = analyzerConfigDocumentStates.Select(a => a.GetAnalyzerConfigAsync(cancellationToken));
+                    var analyzerConfigs = await Task.WhenAll(tasks).ConfigureAwait(false);
+
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    return AnalyzerConfigSet.Create(analyzerConfigs);
+                },
+                synchronousComputeFunction: cancellationToken =>
+                {
+                    var analyzerConfigs = analyzerConfigDocumentStates.SelectAsArray(a => a.GetAnalyzerConfig(cancellationToken));
+                    return AnalyzerConfigSet.Create(analyzerConfigs);
+                },
+                cacheResult: true);
```

Comment:
note: i'm not relaly sure what all the above types are really for.

Reply:
How can I help clarify?